### PR TITLE
Linux Kernel Selftest from linux source

### PR DESCRIPTION
**Pull request by abdhaleegit**
Download linux source repository for the given git/http/tar
url, compile and run the kernel self test.

Signed-off-by: Abdul Haleem <abdhalee@linux.vnet.ibm.com>

### CODE REVIEW

File: kernel/kselftest.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+            version = self.params.get('version', default='4.8.6')
+            tarball_base = 'linux-%s.tar.gz' % (version)
+            tarball_url = '%s/v%s.x/%s' % (url, version[:1], tarball_base)
+            self.log.info('Downloading linux kernel tarball')
+            self.tarball = self.fetch_asset(tarball_url)
+            archive.extract(self.tarball, self.srcdir)
+            self.srcdir = self.srcdir + '/linux-%s' % (version)
+        else:
+            self.log.info('Cloning linux kernel source')
+            git.get_repo(url, destination_dir=self.srcdir)
```
Comment (PraveenPenguin) [inline kernel/kselftest.py]: can kernel avocado API be used ?

```
kern = kernel.KernelBuild(
            version, config_path=config_path, work_dir=None, data_dirs=None)
# Seeting the default kernel URL
kern.URL = 'https://www.kernel.org/pub/linux/kernel/v4.x/'
# download the kernel tarball from kernel.org
kern.download()
# Uncompress the kernel archive to the work directory
kern.uncompress()
```
version can be `4.8.6`

File: kernel/kselftest.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+        url = self.params.get(
+            'url', default='https://www.kernel.org/pub/linux/kernel')
+        if 'git' not in url:
+            version = self.params.get('version', default='4.8.6')
+            tarball_base = 'linux-%s.tar.gz' % (version)
+            tarball_url = '%s/v%s.x/%s' % (url, version[:1], tarball_base)
+            self.log.info('Downloading linux kernel tarball')
+            self.tarball = self.fetch_asset(tarball_url)
+            archive.extract(self.tarball, self.srcdir)
+            self.srcdir = self.srcdir + '/linux-%s' % (version)
```
Comment (PraveenPenguin) [inline kernel/kselftest.py]: please use os.path.join for concatenate directory

File: kernel/kselftest.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+            self.log.info('Cloning linux kernel source')
+            git.get_repo(url, destination_dir=self.srcdir)
+
+        build.make(self.srcdir + self.testdir)
+
+    def test(self):
+        """
+        Execute the kernel selftest
+        """
+        build.make(self.srcdir + self.testdir, extra_args='run_tests')
```
Comment (PraveenPenguin) [inline kernel/kselftest.py]: here as well please use `os.path.join`

File: kernel/kselftest.py.data/README [hunk 4]
```diff
@@ -0,0 +1,3 @@
+This program runs the selftest available with the linux kernel source,
+the script takes input as tarball url with version or a git repository link
+of linux source, compile and run the tests available at / tools / testing / selftests
```
Comment (PraveenPenguin) [inline kernel/kselftest.py.data/README]: please remove the white space in` / tools / testing / selftests  `

File: kernel/kselftest.py (hunk 1 above)
Comment (PraveenPenguin) [inline kernel/kselftest.py]: @abdhaleegit oh yes calling build.make() make sense , but not get this point of URL , you can set URL as i mentioned 
`kern.URL = 'https://www.kernel.org/pub/linux/kernel/v4.x/'` using this you can over write class variable 
and why git url is needed ?
for build.make i think kernel.py can be used like this way 

```
version = self.params.get('version', default='4.8.6')
kern = kernel.KernelBuild(
            version, config_path=config_path, work_dir=self.srcdir, data_dirs=None)
# Seeting the default kernel URL
kern.URL = self.params.get('kernel_version', default='https://www.kernel.org/pub/linux/kernel/v4.x/')

# download the kernel tarball from kernel.org
kern.download()
# Uncompress the kernel archive to the work directory
kern.uncompress()
build.make(os.path.join(self.srcdir , self.testdir))
```

Comment (PraveenPenguin) [inline kernel/kselftest.py]: one more edition you need to again redefine `self.srcdir` variable 
`self.srcdir = os.path.join(self.srcdir , 'linux-%s' % (version))`

File: kernel/kselftest.py (hunk 3 above)
Comment (PraveenPenguin) [inline kernel/kselftest.py]: @abdhaleegit is there any fail condition thinking when self test fail we can notify user about that , 

is yes , build.run_make() can be used in that case

Comment (abdhaleegit) [inline kernel/kselftest.py]: the stdout of the selftest displays PASS, FAIL, OK, ERROR, FAILED all these for each tests. it is transparant to user, we can fail self test if test FAIL/ERROR/FAILED, can this be done by build.run_make() ?

Comment (abdhaleegit) [inline kernel/kselftest.py]: thanks

**Pull request closed without merging**